src/extension/index.ts

```diff
@@ -0,0 +1,193 @@
+import { Baser } from './baser';
+import { loadLogs } from './loadLogs';
+import { Panel } from './panel';
+import { regionToSelection } from './regionToSelection';
+import { ResultDiagnostic } from './resultDiagnostic';
+import { Store } from './store';
+import { update, updateChannelConfigSection } from './update';
+
+export async function activate(context: ExtensionContext) {
+    const disposables = context.subscriptions;
+    Store.extensionPath = context.extensionPath;
+    Store.globalState = context.globalState;
+    disposables.push(commands.registerCommand('sarif.clearState', () => {
+        context.globalState.update('view', undefined);
+        commands.executeCommand('workbench.action.reloadWindow');
+    }));
+    const store = new Store();
+
+    // Basing
+    const urisNonSarif = await workspace.findFiles('**/*', '.sarif'); // Ignore folders?
+    const fileAndUris = urisNonSarif.map(uri => [uri.path.split('/').pop(), uri.path]) as [string, string][];
+    const baser = new Baser(mapDistinct(fileAndUris), store);
+
+    // Panel
+    const panel = new Panel(context, baser, store);
+    disposables.push(commands.registerCommand('sarif.showPanel', () => panel.show()));
+
+    // Diagnostics
+    const diagsAll = languages.createDiagnosticCollection('SARIF');
+    const setDiags = (doc: TextDocument) => {
+        if (doc.fileName.endsWith('.git')) return;
+        const artifactPath = baser.translateLocalToArtifact(doc.uri.path);
+        const severities = {
+            error: DiagnosticSeverity.Error,
+            warning: DiagnosticSeverity.Warning,
+        } as Record<string, DiagnosticSeverity>;
+        const diags = store.results
+            .filter(result => result._uri === artifactPath)
+            .map(result => new ResultDiagnostic(
+                regionToSelection(doc, result._region),
+                result._message ?? '—',
+                severities[result.level ?? ''] ?? DiagnosticSeverity.Information, // note, none, undefined.
+                result,
+            ));
+        diagsAll.set(doc.uri, diags);
+    };
+    workspace.textDocuments.forEach(setDiags);
+    workspace.onDidOpenTextDocument(setDiags);
+    workspace.onDidCloseTextDocument(doc => diagsAll.delete(doc.uri)); // Spurious *.git deletes don't hurt.
+    observe(store.logs, () => workspace.textDocuments.forEach(setDiags));
+
+    // Open Documents <-sync-> Store.logs
+    const syncActiveLog = async (doc: TextDocument) => {
+        if (!doc.fileName.match(/\.sarif$/i)) return;
+        if (store.logs.some(log => log._uri === doc.uri.toString())) return;
+        store.logs.push(...await loadLogs([doc.uri]));
+        panel.show();
+    };
+    workspace.textDocuments.forEach(syncActiveLog);
+    workspace.onDidOpenTextDocument(syncActiveLog);
+    workspace.onDidCloseTextDocument(doc => {
+        if (!doc.fileName.match(/\.sarif$/i)) return;
+        store.logs.removeWhere(log => log._uri === doc.uri.toString());
+    });
```

Comment:
nit: could just combine with the call backs above?

Reply:
One is for push(), the other is for removeWhere(). How are you proposing to merge?

---
In reply to: [441203639](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441203639) [](ancestors = 441203639)

Reply:
I was thinking the callbacks to `workspace.textDocuments.forEach` and `workspace.onDidOpenTextDocument` and `workspace.onDidCloseTextDocument` could be combined, but this is fine too 

---
In reply to: [441205991](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441205991) [](ancestors = 441205991,441203639)

Reply:
As in, combined with the ones from `activateDiagnostics` 

---
In reply to: [442377153](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r442377153) [](ancestors = 442377153,441205991,441203639)

Reply:
Ah, if we merged it, then by the time the handler is called, we wouldn't know if the `doc` was being closed or opened.

---
In reply to: [442377664](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r442377664) [](ancestors = 442377664,442377153,441205991,441203639)

Reply:
I think `syncActiveLog` might be misleading and I have renamed it to `addLog`.

---
In reply to: [442555269](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r442555269) [](ancestors = 442555269,442377664,442377153,441205991,441203639)